src/devices/machine/gt913.cpp

```diff
@@ -0,0 +1,569 @@
+
+	recompute_bcount(event_time);
+}
+
+void gt913_device::execute_set_input(int inputnum, int state)
+{
+	if (inputnum == INPUT_LINE_NMI)
+	{
+		if (state == ASSERT_LINE && m_nmi_status != state)
+			m_nmi_pending = true;
+		state = m_nmi_status;
+	}
+	else if (inputnum <= INPUT_LINE_IRQ7)
+	{
+		if (state)
+			m_irq_pending |= (1 << inputnum);
+		else
+			m_irq_pending &= ~(1 << inputnum);
+	}
+	update_irq_filter();
+}
+
+void gt913_device::set_irq_enable(int inputnum, int state)
+{
+	if (state)
+		m_irq_enable |= (1 << inputnum);
+	else
+		m_irq_enable &= ~(1 << inputnum);
+
+	update_irq_filter();
+}
+
+void gt913_device::device_start()
+{
+	h8_device::device_start();
+
+	m_bank->configure_entries(0, m_rom->bytes() >> 12, m_rom->base(), 1 << 12);
+
+	m_timer[0] = timer_alloc(0);
+	m_timer[1] = timer_alloc(1);
+
+	decode_opcodes();
+
+	save_item(NAME(m_banknum));
+
+	save_item(NAME(m_nmi_pending));
+	save_item(NAME(m_nmi_status));
+	save_item(NAME(m_irq_pending));
+	save_item(NAME(m_irq_enable));
+
+	save_item(NAME(m_timer_control));
+	save_item(NAME(m_timer_rate));
+	save_item(NAME(m_uart_control));
+	save_item(NAME(m_adc_enable));
+	save_item(NAME(m_adc_channel));
+	save_item(NAME(m_adc_data));
+	save_item(NAME(m_port_ddr));
+}
+
+void gt913_device::device_reset()
+{
+	h8_device::device_reset();
+
+	m_banknum = 0;
+
+	m_nmi_pending = false;
+	m_nmi_status = CLEAR_LINE;
+	m_irq_pending = 0x00;
+	m_irq_enable = 0xff;
+
+	m_timer_control[0] = m_timer_control[1] = 0x00;
+	m_timer_rate[0] = m_timer_rate[1] = 0;
+
+	m_uart_control = 0x00;
+
+	m_adc_enable = false;
+	m_adc_channel = false;
+	m_adc_data[0] = m_adc_data[1] = 0;
+	m_port_ddr[0] = m_port_ddr[1] = 0;
+}
+
+void gt913_device::device_reset_after_children()
+{
+	h8_device::device_reset_after_children();
+
+	// MIDI UART, asynchronous 8/n/1
+	m_uart->write_xr(0);
+	m_uart->write_xr(1);
+	m_uart->write_swe(0);
+	m_uart->write_nb1(1);
+	m_uart->write_nb2(1);
+	m_uart->write_np(1);
+	m_uart->write_tsb(0);
+	m_uart->write_cs(1);
+}
+
+void gt913_device::do_exec_full()
+{
+	if ((inst_state & 0xfffb80) == 0x0380)
+	{
+		set_bank_num();
+		if (icount <= bcount) { inst_substate = 1; return; }
+		prefetch();
+	}
+	else
+	{
+		h8_device::do_exec_full();
+	}
+}
+
+void gt913_device::do_exec_partial()
+{
+	if ((inst_state & 0xfffb80) == 0x0380)
+	{
+		switch (inst_substate) {
+		case 0:
+			set_bank_num();
+			if (icount <= bcount) { inst_substate = 1; return; }
+			[[fallthrough]];
+		case 1:;
+			prefetch();
+			break;
+		}
+		inst_substate = 0;
+	}
+	else
+	{
+		h8_device::do_exec_partial();
+	}
+}
+
+void gt913_device::set_bank_num()
+{
+	if (BIT(IR[0], 10))
+		TMP1 = IR[0] & 0xf;
+	else
+		TMP1 = r8_r(IR[0]);
+
+	if (BIT(IR[0], 6))
+		m_banknum = (m_banknum & 0xfffc) | (TMP1 & 0x3);
+	else
+		m_banknum = (TMP1 << 2) | (m_banknum & 0x3);
+
+	m_bank->set_entry(m_banknum);
+}
+
+void gt913_device::decode_opcodes()
+{
+	auto rombase = &m_rom->as_u16();
+	const auto size = 0x8000 / 2;
```

Comment:
This is making huge assumptions about the rest of the system, and it’s the wrong place to do something like this.  Is it not possible to work with the H8 CPU code generation to build a core with appropriate decoding?  The 6502 family has variants with different instruction decoding.

Reply:
I agree that's the better way to go, but my main concern was not wanting to bloat `h8.lst` with a bunch of duplicated implementations for dozens of different opcodes. Maybe there's a way around that, though.